crates/db/store-sled/src/unified_mmr/db.rs

```diff
@@ -0,0 +1,515 @@
+            },
+        )
+    }
+
+    fn get_node(&self, mmr_id: MmrId, pos: u64) -> DbResult<[u8; 32]> {
+        self.get_mmr_node(&mmr_id, pos)
+    }
+
+    fn mmr_size(&self, mmr_id: MmrId) -> DbResult<u64> {
+        self.ensure_mmr_metadata(&mmr_id)?;
+        let metadata = self.load_mmr_metadata(&mmr_id)?;
+        Ok(metadata.mmr_size)
+    }
+
+    fn num_leaves(&self, mmr_id: MmrId) -> DbResult<u64> {
+        self.ensure_mmr_metadata(&mmr_id)?;
+        let metadata = self.load_mmr_metadata(&mmr_id)?;
+        Ok(metadata.num_leaves)
+    }
+
+    fn peak_roots(&self, mmr_id: MmrId) -> Vec<[u8; 32]> {
+        self.load_mmr_metadata(&mmr_id)
+            .map(|m| m.peak_roots.into_iter().map(|b| b.0).collect())
+            .unwrap_or_default()
+    }
+
+    fn to_compact(&self, mmr_id: MmrId) -> CompactMmr64 {
+        let metadata = self
+            .load_mmr_metadata(&mmr_id)
+            .unwrap_or_else(|_| MmrMetadata::empty());
+
+        let roots_vec: Vec<_> = metadata
+            .peak_roots
+            .iter()
+            .map(|buf| FixedBytes::<32>::from(buf.0))
+            .collect();
+
+        CompactMmr64 {
+            entries: metadata.num_leaves,
+            cap_log2: 64,
+            roots: roots_vec.into(),
+        }
+    }
+
+    fn pop_leaf(&self, mmr_id: MmrId) -> DbResult<Option<[u8; 32]>> {
+        self.ensure_mmr_metadata(&mmr_id)?;
+
+        self.config.with_retry(
+            (&self.node_tree, &self.meta_tree, &self.hash_index_tree),
+            |(nt, mt, hit)| {
+                let metadata = mt
+                    .get(&mmr_id)?
+                    .expect("MMR metadata must exist after ensure_mmr_metadata");
+
+                let result = MmrAlgorithm::pop_leaf(&metadata, |pos| {
+                    nt.get(&(mmr_id.clone(), pos))?
+                        .map(|buf| buf.0)
+                        .ok_or_else(|| DbError::MmrNodeNotFound(pos, mmr_id.clone()))
+                })
+                .map_err(typed_sled::error::Error::abort)?;
+
+                let Some(result) = result else {
+                    return Ok(None);
+                };
+
+                for pos in &result.nodes_to_remove {
+                    nt.remove(&(mmr_id.clone(), *pos))?;
+                }
+
+                mt.insert(&mmr_id, &result.new_metadata)?;
+
+                // Remove hash -> position mapping for the popped leaf
+                hit.remove(&(mmr_id.clone(), Buf32(result.leaf_hash)))?;
+
+                Ok(Some(result.leaf_hash))
+            },
+        )
+    }
+
+    fn append_leaf_with_preimage(
+        &self,
+        mmr_id: MmrId,
+        hash: [u8; 32],
+        preimage: Vec<u8>,
+    ) -> DbResult<u64> {
+        self.ensure_mmr_metadata(&mmr_id)?;
+
+        self.config.with_retry(
+            (
+                &self.node_tree,
+                &self.meta_tree,
+                &self.hash_index_tree,
+                &self.preimage_tree,
```

Comment:
given we have `append_leaf`, that doesn't take a preimage, this means that `preimage_tree` may not include items for which we actually we do a have leaf.
1. I guess it's fine, but want to double check. Perhaps in implementation we only call this function(?)
2. shouldn't we at least update it to indicate that a leaf exists but we don't know the preimage? say with _None_, if it makes sense, or some distinguished value?

Reply:
I don't understand fully what you mean here. How this would work is that where we use the mmr db, we would use either `append_leaf` or `append_leaf_with_preimage` based on the usecase.

Reply:
Hmm. I don't think that's necessary because `get_preimage` returns `None` if the preimage does not exist. If we try to explicitly insert something(null value like you suggested) then we would just use up the disk without much benefit.